### PR TITLE
Optimize interpolateTypeVars

### PR DESCRIPTION
**Pull request by odersky**
There seemed to have been a large performance regression due to its recent changes. Let's find
out whether that's really the case.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
I wonder what's going on with the benchmarks? We see a lot more volatility than wee used to. This PR seems to improve the situation somewhat for Dotty compilation, but we still have no clue what caused the dramatic degradation in #6301 - it does not look that the changes in this commit were to blame alone.

**Comment by odersky** [comment]
Test performance please

**Comment by smarter** [comment]
I think the bugfix is revealing some underlying bug, there's only very few cases when compiling Dotty there's only a few cases where the old check returned false and the new returned true, but a couple of those have a crazy amount of type variables:
```scala
state: TS[7389, 7388, 7387]
state.ownedVars: 28
locked: 28

state: TS[7389, 7388, 7387]
state.ownedVars: 344
locked: 344

state: TS[7389, 7388, 7387]
state.ownedVars: 344
locked: 344

state: TS[7389, 7388, 7387]
state.ownedVars: 344
locked: 345
```

**Comment by odersky** [comment]
Would be interesting to find out for what code these large sets are created.

**Comment by smarter** [comment]
It looks like all these type variables are created using `newDepTypeVar`, I'm not sure how these are supposed to interact with interpolation.

**Comment by odersky** [comment]
> It looks like all these type variables are created using newDepTypeVar, I'm not sure how these are supposed to interact with interpolation.

They get added to typerState.ownedVars and subsequently `typedArg` or implicit search will move them to `locked`. A question is whether we can garbage collect them more aggressively. I am not sure.

**Comment by odersky** [comment]
> It looks like all these type variables are created using newDepTypeVar, I'm not sure how these are supposed to interact with interpolation.

Yes, but what is the source file that causes so many dependent TypeVars to be created?

**Comment by smarter** [comment]
Most seem to be either in `Trees.scala` or `untpd.scala` (apparently because of `def postProcess`)

**Comment by odersky** [comment]
That makes sense. `postprocess` (and `finalize`) are dependent methods. The question why the typevars are not collected sooner is still open.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
OK, we are back to normal. We could still check whether we have to create so many ownedVars in the first place.

**Comment by odersky** [comment]
I propose to merge this now to get back a stable baseline for other performance tests.

File: compiler/src/dotty/tools/dotc/util/SimpleIdentitySet.scala [hunk 1]
```diff
@@ -115,5 +139,78 @@ object SimpleIdentitySet {
...
+          val thisSize = this.size
+          val thatSize = that.size
+          val thatElems = that.xs
+          var i = 0
+          var searchStart = 0
+          while (i < thisSize) {
+            val elem = this.xs(i)
+            var j = searchStart    // search thatElems in round robin fashion, starting one after latest hit
+            var missing = false
+            while (!missing && elem != thatElems(j)) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/util/SimpleIdentitySet.scala]:
```suggestion
            while (!missing && elem `ne` thatElems(j)) {
```

File: compiler/src/dotty/tools/dotc/util/SimpleIdentitySet.scala [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
+  def exists[E >: Elem <: AnyRef](p: E => Boolean): Boolean
+  def /: [A, E >: Elem <: AnyRef](z: A)(f: (A, E) => A): A
+  def toList: List[Elem]
   def ++ [E >: Elem <: AnyRef](that: SimpleIdentitySet[E]): SimpleIdentitySet[E] =
-    ((this: SimpleIdentitySet[E]) /: that.toList)(_ + _)
-  def -- [E >: Elem <: AnyRef](that: SimpleIdentitySet[E]): SimpleIdentitySet[Elem] =
-    (this /: that.toList)(_ - _)
+    if (this.size == 0) that
+    else if (that.size == 0) this
+    else concat(that)
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/util/SimpleIdentitySet.scala]: I originally thought of putting concat in Set 1-3, so wanted to factor put the common behavior. But the way it is now a single method is better.

**Comment by odersky** [comment]
> There's another use of % below for searchStart, although that one is in a less tight loop.

Yes, I saw that. I don't think we need to optimize it since it's in the outer loop.

**Pull request merged**